src/rulesets/Base/mapreduce.jl

```diff
@@ -187,6 +187,55 @@ function rrule(::typeof(cumsum), x::AbstractArray; dims::Integer)
 end
 rrule(::typeof(cumsum), x::AbstractVector) = rrule(cumsum, x; dims=1)
 
+#####
+##### `maximum(f, xs)`, `minimum(f, xs)`
+#####
+
+# Rules for `maximum(x)` live with `findmax(x)` in array.jl
+
+for mimum in (:minimum, :maximum)
+    findm = Symbol(:find, string(mimum)[1:3])
+
+    @eval function rrule(
+        config::RuleConfig{>:HasReverseMode},
+        ::typeof($mimum),
+        f::F,
+        xs::AbstractArray{<:Number};
+        dims = :,
+    ) where {F}
+        project = ProjectTo(xs)
+        if dims isa Colon && VERSION >= v"1.7"
+            # The easy case is when we can use `findmax` to get index, and write into it:
+            y, ind = $findm(f, xs)
+            function minormax_f_back1(dy)
+                # Notice this evaluates `f` one more time, but this shouldn't matter unless `f` is
+                # sateful, in which case both this and `maximum(f.(xs))` give uncertain results.
+                _, one_back = rrule_via_ad(config, f, xs[ind])
```

Comment:
(What's changed in this suggestion?)

Reply:
```suggestion
                y_ad, one_back = rrule_via_ad(config, f, xs[ind])
                y ≈ y_ad || throw(ArgumentError("expected `f` to give same result with AD, got $y_ad != $y"))
```

Reply:
Now an error. `≈` takes a few ns, since this must be a scalar, so should be fine.

Reply:
This test breaks type inference. I don't know why.

Reply:
hmm, weird, this is the MWE I think
```julia
julia> myfunc(x) = x > 0 ? 0.0 : throw(DomainError("smaller"))
myfunc (generic function with 1 method)

julia> @code_warntype myfunc(2.0)
MethodInstance for myfunc(::Float64)
  from myfunc(x) in Main at REPL[3]:1
Arguments
  #self#::Core.Const(myfunc)
  x::Float64
Body::Float64
1 ─ %1 = (x > 0)::Bool
└──      goto #3 if not %1
2 ─      return 0.0
3 ─ %4 = Main.DomainError("smaller")::Any
│        Main.throw(%4)
└──      Core.Const(:(return %5))
```
is this expected or a julia bug?

Reply:
Making dims positional not kw, it's more like this: 
```
julia> @code_warntype rrule(CFG, maximum, abs, [1,2,3], :)
MethodInstance for ChainRulesCore.rrule(::ChainRulesTestUtils.TestConfig, ::typeof(maximum), ::typeof(abs), ::Vector{Int64}, ::Colon)
  from rrule(config::RuleConfig{>:HasReverseMode}, ::typeof(maximum), f::F, xs::AbstractArray{<:Number}, dims) where F in ChainRules at /Users/me/.julia/dev/ChainRules/src/rulesets/Base/mapreduce.jl:199
Static Parameters
  F = typeof(abs)
Arguments
  #self#::Core.Const(ChainRulesCore.rrule)
  config::ChainRulesTestUtils.TestConfig
  @_3::Core.Const(maximum)
  f::Core.Const(abs)
...
  y@_16::Core.Box
  project::ProjectTo{AbstractArray, NamedTuple{(:element, :axes), Tuple{ProjectTo{Float64, NamedTuple{(), Tuple{}}}, Tuple{Base.OneTo{Int64}}}}}
  y@_18::Union{}
  y@_19::Union{}
Body::Tuple{Any, ChainRules.var"#minormax_f_back1#2045"{ChainRulesTestUtils.TestConfig, typeof(abs), Vector{Int64}, Colon, Int64, ProjectTo{AbstractArray, NamedTuple{(:element, :axes), Tuple{ProjectTo{Float64, NamedTuple{(), Tuple{}}}, Tuple{Base.OneTo{Int64}}}}}}}
```
Core.Box from closing over y, maybe it can't prove that `isapprox(y_ad, y)` won't change `y`?

Reply:
Solved below, by changing the name of y.